CDS/src/org/icpc/tools/cds/presentations/Client.java

```diff
@@ -258,27 +265,27 @@ protected long handlePing() throws IOException {
 
 		Trace.trace(Trace.INFO, Integer.toHexString(uid) + " - Recent pings: [" + sb.toString() + "] ");
 
-		// if delta difference is < 20ms, we're close enough
+		// delta difference is < 20ms for at least 2 pings, we're close enough to sync time
 		if (count > 1 && max < 20) {
 			writeTime(total / count);
-			return 0;
+			return true;
 		}
 
 		if (count < 3) {
-			// not enough response time data yet, ping again in 500ms
+			// not enough response time data yet, ping again
 			writePing();
-			return 500;
+			return false;
 		}
 
 		if (count == 3 && max - min > 50) {
-			// very inconsistent response times, ping once more in 750ms
+			// very inconsistent response times, ping at least once more
 			writePing();
-			return 750;
+			return false;
 		}
 
-		// get rid of 'worst' time outlier and determine new average time sync
+		// we've had at least 3 pings, get rid of 'worst' time outlier and send average time sync
 		writeTime((total - max) / (count - 1));
```

Comment:
ideally, the client currentTimeMillis should be communicated, so cds could send a time delta value to the client based on the shortest pings. with the current writeTime, it is very sensitive to the delay of _that_ call.

Reply:
(not a big problems as long as pings and requests take milliseconds)